packages/terra-pills/src/PillList.jsx

```diff
@@ -0,0 +1,186 @@
+
+const cx = classNamesBind.bind(styles);
+
+const propTypes = {
+  /**
+   * The string that labels the collection of pills, used in cased where the text label is not visible on the screen and required for minimum accessibility standards.
+   * Providing this prop adds the `aria-label` attribute to the Pill List container element. (Required)
+   */
+  ariaLabel: PropTypes.string.isRequired,
+  /**
+   * If a visible text label is used with the collection of pills, provide a string of the ID for the html element containing the visible text label.
+   * Supplying the 'ariaLabelledBy' prop will override the 'ariaLabel' prop and adds the `aria-labelledby` attribute instead to the Pill List container element.
+   *
+   * ![IMPORTANT](https://badgen.net/badge/UX/Accessibility/blue) For best practices, ensure the visible text in the html element `id` provided to the Pill List
+   * `ariaLabelledby`prop matches the text provided to the `ariaLabel` prop, for consistency in the case of fallback or errors.
+   */
+  ariaLabelledBy: PropTypes.string,
+  /**
+   * If additional visible information text is used, provide a string containing the IDs for html elements that help describe the intent of the pills and the Pill List container.
+   */
+  ariaDescribedBy: PropTypes.string,
+  /**
+   * The content to be shown in the pills container. Should only contain the sub-component pills.
+   */
+  children: PropTypes.node,
+  /**
+   * Determines if the Pill List is rolled up or not.
+   */
+  isCollapsed: PropTypes.bool,
+  /**
+   * Callback function triggered on click/key press of the roll-up pill
+   */
+  rollUpPillOnTrigger: PropTypes.func,
+  /**
+   * @private
+   * The intl object to be injected for translations.
+   */
+  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
+};
+
+const defaultProps = {
+  ariaLabelledBy: undefined,
+  ariaDescribedBy: undefined,
+  children: undefined,
+  isCollapsed: false,
+  rollUpPillOnTrigger: undefined,
+};
+
+const PillList = (props) => {
+  const {
+    ariaLabel,
+    ariaLabelledBy,
+    ariaDescribedBy,
+    children,
+    isCollapsed,
+    rollUpPillOnTrigger,
+    intl,
+    ...customProps
+  } = props;
+
+  const theme = React.useContext(ThemeContext);
+  const pillListRef = useRef();
+  const rollUpPillRef = useRef();
+  const [rollUpCount, setRollUpCount] = useState(0);
+
+  const handleRollUp = () => {
+    rollUpPillRef.current.removeAttribute('style');
+    let pillListHeight = pillListRef.current.clientHeight;
+    const rollUpPillStyles = window.getComputedStyle(rollUpPillRef.current);
+    const pillHeight = rollUpPillRef.current.offsetHeight + Math.round(parseFloat(rollUpPillStyles.marginTop)) + Math.round(parseFloat(rollUpPillStyles.marginBottom));
+
+    if (pillListHeight <= pillHeight) {
+      rollUpPillRef.current.style.display = 'none';
+      return;
+    }
+
+    const childPills = pillListRef.current.querySelectorAll('[class*="pill-container"]');
```

Comment:
Classnames may not always be query-able(?) depending on how the css modules are processed. I'd query based on a static data attribute or grab a ref to it somehow.

Reply:
Correct me if I'm wrong, but this logic looks like it would loop forever if there was ever a non-Pill child provided that was larger than the pillHeight? Is that a scenario we need to support? 

If not, we should refactor the PillList to retain refs to each Pill (or a container around them) and simply loop over them to hide/show them, cutting out this math-heavy algorithm that is boils down to "hide (or show) all non-rollup pills".

If we do need to support non-Pill children, I don't think we can reliably roll up in this way.

Reply:
1)I will include a data-attribute and query it on the same.

2)According to the props description, we are restricting the children to be just Pills.


3)I'm sorry but I did not get this part.
` How do we handle scenarios where the PillList is sized smaller after performing the rollup? `

Reply:
data-attribute included in this commit and queried on the same 1319a8e97ab1b290bf12be47dce185d182b81190

Reply:
Regarding 2, I think propType enforcement would be sufficient normally, but that `while` loop (and the potential for the infinite loop) makes me nervous. Got a couple options:

A. Throw exceptions if non-Pill children are provided (clarity at the expense of flexibility)
B. Filter out non-Pill children and never render them (flexibility at the expense of clarity)
C. Switch way from a `while` loop to a `for` loop with `break`s so that we have a defined maximum loop count

I'd recommend A _and_ C.

Regarding 3, I'm referring to a scenario when a rollup has occurred, but then the PillList itself is shrunk. Currently, it will result in the Pills wrapping like normal, with the RollUpPill being moved to subsequent lines. I am unsure if the requirement is to ensure rolled-up pills are always contained to a single row, or if we're okay with that subsequent wrapping occurring. Here's a gif I took of one of the tests I manually manipulated:

![pill_rollup](https://user-images.githubusercontent.com/1711637/125695405-ef3c9866-40dd-4ad8-91aa-db852e9a5d6a.gif)